### PR TITLE
Fix Login form validation

### PR DESCRIPTION
**Pull request by josemigallas**
Sorry, I did it again 😅😅😅😅😅

But these pages needed cleaning up/fixing/refactoring and it was the perfect opportunity.

[THREESCALE-9760: Login screen: "Sign in" button dissabled after first login attempt](https://issues.redhat.com/browse/THREESCALE-9760)

### Signup
https://github.com/3scale/porta/assets/11672286/8ea6a18d-008a-4573-ba09-a49354f90464

### Change password
https://github.com/3scale/porta/assets/11672286/07045a20-65bc-47c7-a596-0e7047e707f5

### Request password
https://github.com/3scale/porta/assets/11672286/420a8437-5a8e-456b-96ff-a6737de73e46

### Log in
https://github.com/3scale/porta/assets/11672286/c7b24535-e22a-4336-a459-31ca397b15a3

### CODE REVIEW

**Review by akostadinov — APPROVED** [review]
tl;dr but video looks good :)

File: app/javascript/src/Login/components/ChangePasswordPage.tsx [hunk 1]
```diff
@@ -64,28 +61,75 @@ const ChangePassword: FunctionComponent<Props> = ({
...
+        <input name="utf8" type="hidden" value="✓" />
+        <CSRFToken />
+
+        <FormGroup
+          isRequired
+          autoComplete="off"
+          fieldId="user_password"
+          helperTextInvalid={validation?.password?.[0]}
+          label="Password"
+          validated={(validationVisibility.password && validation?.password) ? 'error' : 'default'}
```
Comment (josemigallas) [inline app/javascript/src/Login/components/ChangePasswordPage.tsx:87]: DRY move these validation values into const

File: app/javascript/src/Login/components/LoginForm.tsx [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+  error,
+  providerSessionsPath,
+  session
+}) => {
+  const [username, setUsername] = useState(session.username ?? '')
+  const [password, setPassword] = useState('')
+  const [validationVisibility, setValidationVisibility] = useState({
+    username: false,
+    password: false
+  })
```
Comment (josemigallas) [inline app/javascript/src/Login/components/LoginForm.tsx:36]: Would it worth it combining `useState`s and validationVisibility into a custom hook?
```
  const [state, setState, showError, setShowError] = useValidatedFormState({
    username: session.username ?? '',
    password: ''
  })

File: app/javascript/src/Login/components/LoginForm.tsx [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+  const onUsernameChange = (value: string) => {
+    setUsername(value)
+    setValidationVisibility(prev => ({ ...prev, username: false }))
+  }
+
+  const onPasswordChange = (value: string) => {
+    setPassword(value)
+    setValidationVisibility(prev => ({ ...prev, password: false }))
+  }
```
Comment (josemigallas) [inline app/javascript/src/Login/components/LoginForm.tsx:46]: Somehow move this into custom hook... create an object of `onChange`s and another of `onBlur`s?

**Review by jlledom — CHANGES_REQUESTED** [review]
A few issues I found:

- Information messages have lost their styling after:
  - Signing up
  - Restoring password
  - Requesting a new password
This is how it looked before:
image
- I can only compile assets locally with the `webpack:dev` task. When I try to run `bundle exec rails assets:clobber assets:precompile` I get this error:

```
Entrypoint wizard_backend_api = js/wizard_backend_api-d9a126d4d8a95916d4fb.js js/wizard_backend_api-d9a126d4d8a95916d4fb.js.map
 [958] ./app/javascript/packs/active_docs.ts 663 bytes {6} [built]
 [980] ./app/javascript/packs/braintree_provider_form.ts 1.37 KiB {13} [built]
 [985] multi ./app/javascript/packs/codemirror.scss ./app/javascript/packs/codemirror.ts 40 bytes {16} [built]
 [993] multi ./app/javascript/packs/dashboard.scss ./app/javascript/packs/dashboard.ts 40 bytes {17} [built]
[1004] multi ./app/javascript/packs/inline_chart.scss ./app/javascript/packs/inline_chart.tsx 40 bytes {21} [built]
[1006] ./app/javascript/packs/invoice_payment.ts 3.29 KiB {22} [built]
[1007] multi ./app/javascript/packs/login.scss ./app/javascript/packs/login.ts 40 bytes {23} [built]
[1096] ./app/javascript/packs/plans_metrics.ts 934 bytes {31} [built]
[1488] multi ./app/javascript/packs/provider.scss ./app/javascript/packs/provider.ts 40 bytes {36} [built]
[1594] ./app/javascript/packs/service_active_docs.ts 1.58 KiB {40} [built]
[1597] ./app/javascript/packs/validate_signup.ts 1.66 KiB {46} [built]
[1598] ./app/javascript/packs/wizard_backend_api.ts 646 bytes {52} [built]
[1599] ./app/javascript/packs/policies.ts + 80 modules 168 KiB {32} [built]
       | ./app/javascript/packs/policies.ts 515 bytes [built]
       | ./app/javascript/src/Policies/index.tsx 620 bytes [built]
       | ./app/javascript/src/Policies/actions/index.ts 264 bytes [built]
       | ./app/javascript/src/Policies/reducers/initialState.ts 301 bytes [built]
       | ./app/javascript/src/Policies/store/configureStore.ts 1.51 KiB [built]
       | ./app/javascript/src/Policies/actions/OriginalPolicyChain.ts 98 bytes [built]
       | ./app/javascript/src/Policies/actions/PolicyChain.ts 1.02 KiB [built]
       | ./app/javascript/src/Policies/actions/PolicyConfig.ts 87 bytes [built]
       | ./app/javascript/src/Policies/actions/PolicyRegistry.ts 423 bytes [built]
       | ./app/javascript/src/Policies/actions/UISettings.ts 288 bytes [built]
       | ./app/javascript/src/Policies/actions/ThunkActions.ts 2.93 KiB [built]
       | ./app/javascript/src/Policies/middleware/PolicyChain.ts 2 KiB [built]
       | ./app/javascript/src/Policies/reducers/index.ts 546 bytes [built]
       | ./app/javascript/src/Policies/util.ts 1.16 KiB [built]
       | ./app/javascript/src/Policies/reducers/UISettings.ts 894 bytes [built]
       |     + 66 hidden modules
[1600] ./app/javascript/packs/provider_stats.js + 17 modules 18.1 KiB {38} [built]
       | ./app/javascript/packs/provider_stats.js 676 bytes [built]
       | ./app/javascript/src/Stats/provider/stats_usage.js 1.23 KiB [built]
       | ./app/javascript/src/Stats/provider/stats_days_of_week.js 1.58 KiB [built]
       | ./app/javascript/src/Stats/provider/stats_hours_of_day.js 1.76 KiB [built]
       | ./app/javascript/src/Stats/provider/stats_top_apps.js 2.72 KiB [built]
       | ./app/javascript/src/Stats/provider/stats_application.js 1.17 KiB [built]
       | ./app/javascript/src/Stats/provider/stats_response_codes.js 3.37 KiB [built]
       | ./app/javascript/src/Stats/lib/usage_backend_api_source_collector.js 292 bytes [built]
       | ./app/javascript/src/Stats/lib/usage_source_collector.js 250 bytes [built]
       | ./app/javascript/src/Stats/lib/average_chart_manager.js 224 bytes [built]
       | ./app/javascript/src/Stats/lib/average_metrics_source.js 1.11 KiB [built]
       | ./app/javascript/src/Stats/lib/application_details.js 227 bytes [built]
       | ./app/javascript/src/Stats/lib/applications_table.js 1.26 KiB [built]
       | ./app/javascript/src/Stats/lib/top_apps_source_collector.js 1.02 KiB [built]
       | ./app/javascript/src/Stats/lib/usage_backend_api_metrics_source.js 197 bytes [built]
       |     + 3 hidden modules
[1603] ./app/javascript/packs/new_service.ts + 11 modules 12.8 KiB {28} [built]
       | ./app/javascript/packs/new_service.ts 584 bytes [built]
       |     + 11 hidden modules
    + 2097 hidden modules

ERROR in /home/jlledom/redhat/porta/app/javascript/src/Types/index.ts
[tsl] ERROR in /home/jlledom/redhat/porta/app/javascript/src/Types/index.ts(5,15)
      TS2307: Cannot find module 'Types/Signup' or its corresponding type declarations.


Process finished with exit code 1
```

**Comment by jlledom** [comment]
> ERROR in /home/jlledom/redhat/porta/app/javascript/src/Types/index.ts
> [tsl] ERROR in /home/jlledom/redhat/porta/app/javascript/src/Types/index.ts(5,15)
>       TS2307: Cannot find module 'Types/Signup' or its corresponding type declarations.

Happenend also on CircleCI:

https://app.circleci.com/pipelines/github/3scale/porta/25528/workflows/a7b142b8-5758-4375-8d04-8dfa6fede508/jobs/287545

**Comment by josemigallas** [comment]
@jlledom I left a couple bugs, they should be fixed now.

Regarding the loss in styles: what we have right now are "made up" styles, there is no official way to show a notification, only errors. So, in theory, only errors should be red and have an icon. 

However I see in then design guidelines that Alerts can be used to show messages, and it looks like this (IMHO this is better):
Screenshot 2023-08-21 at 17 37 12
Screenshot 2023-08-21 at 17 34 20
Screenshot 2023-08-21 at 17 34 02
Screenshot 2023-08-21 at 17 33 46

**Comment by akostadinov** [comment]
This "email not found" and "a password reset link has been emailed to you" should in fact show the same informational message in both case, something like:
> if the provided email address belongs to an account, a password reset link should be sent

Not good, too sleepy now. But we can check the message other services show. The idea is that we shouldn't expose the information whether this email is registered or not.

**Review by jlledom — APPROVED** [review]

**Comment by josemigallas** [comment]
> This "email not found" and "a password reset link has been emailed to you" should in fact show the same informational message in both case, something like:
> 
> > if the provided email address belongs to an account, a password reset link should be sent
> 
> Not good, too sleepy now. But we can check the message other services show. The idea is that we shouldn't expose the information whether this email is registered or not.

I totally agree with you. We can check with @dfennessy the proper wording and update it now or later.

**Review by jlledom — APPROVED** [review]
